File: linux/samples/ReEncode/options.cpp
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/
#include <iostream>
#include "options.h"
#include "program_options.h"
#include "util.h"

using namespace std;

void help(primo::program_options::OptionsConfig<char>& optcfg)
{
    cout << "Usage: ReEncode --input inputFile.mp4 --output outputFile.mp4 [--reEncodeAudio yes|no] [--reEncodeVideo yes|no]\n" << endl;
    primo::program_options::doHelp(cout, optcfg);
}

void setDefaultOptions(Options& opt)
{
    opt.inputFile = getExeDir() + "/../assets/mov/big_buck_bunny_trailer_iphone.m4v";
    opt.outputFile = getExeDir() + "/reencode_output.mp4";
}

bool validateOptions(Options& opt)
{
    return (!opt.inputFile.empty() && 
            !opt.outputFile.empty());    
}

ErrorCodes prepareOptions(Options &opt, int argc, char* argv[])
{
    if (argc < 2)
    {
        setDefaultOptions(opt);
        cout << "Using defaults:\n";
        cout << " --input " << opt.inputFile;
        cout << " --output " << opt.outputFile;
        cout << " --reEncodeVideo " << opt.reEncodeVideo;
        cout << " --reEncodeAudio " << opt.reEncodeAudio;
        
        cout << endl;
        return Parsed;
    }
    
    primo::program_options::OptionsConfig<char> optcfg;
    optcfg.addOptions()
            ("help,h", opt.help, "")
            ("input,i", opt.inputFile, string(), "input mp4 file")
            ("output,o", opt.outputFile, string(), "output mp4 file")
            ("reEncodeAudio,a", opt.reEncodeAudio, YesNo(true), "re-encode audio, yes|no")
            ("reEncodeVideo,v", opt.reEncodeVideo, YesNo(true), "re-encode video, yes|no");
       
    try
    {
        primo::program_options::scanArgv(optcfg, argc, argv);
    }
    catch (primo::program_options::ParseFailure<char> &ex)
    {
        cout << ex.message() << endl;
        help(optcfg);
        return Error;
    }
    
    if (opt.help)
    {
        help(optcfg);
        return Command;
    }

    if (!validateOptions(opt))
    {
        help(optcfg);
        return Error;
    }
    
    return Parsed;
}

File: windows/net/samples/ReEncode/Options.cs
/*
 *  Copyright (c) 2013 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/
using System;
using System.IO;
using CommandLine;
using CommandLine.Text;
using PrimoSoftware.AVBlocks;

namespace ReEncodeSample
{
    class Options
    {
        // ReEncodeSample command line options
        [Option('?', "help", HelpText = "Display this help screen")]
        public bool Help { get; set; }

        [Option('i', "input", HelpText = "input file")]
        public string InputFile { get; set; }

        [Option('o', "output", HelpText = "output file")]
        public string OutputFile { get; set; }

        [Option("reEncodeAudio", HelpText = "Re-encode audio forced, yes|no")]
        public string ReEncodeAudioStrValue { get; set; }

        [Option("reEncodeVideo", HelpText = "Re-encode video forced, yes|no")]
        public string ReEncodeVideoStrValue { get; set; }

        // The program parses the command line options and sets these properties
        public bool Error { get; private set; }

        public bool ReEncodeVideo { get { return IsYesNoOptionEnabled(ReEncodeVideoStrValue); } }
        public bool ReEncodeAudio { get { return IsYesNoOptionEnabled(ReEncodeAudioStrValue); } }

        string GetUsage()
        {
            return HelpText.AutoBuild(this, (HelpText current) => HelpText.DefaultParsingErrorsHandler(this, current));
        }

        bool IsYesNoOptionEnabled(string val)
        {
            if (string.IsNullOrEmpty(val))
                return true;

            val = val.ToLowerInvariant();
            return (val == "yes") || (val == "y");
        }

        void PrintUsage()
        {
            Console.WriteLine("Usage: Usage: ReEncode [--input inputFile.mp4] [--output outputFile.mp4] [--reEncodeAudio yes|no] [--reEncodeVideo yes|no]");
            Console.WriteLine(GetUsage());
        }

        void ResetOptions()
        {
            Help = false;
            Error = false;
            InputFile = null;
            OutputFile = null;
            ReEncodeAudioStrValue = null;
            ReEncodeVideoStrValue = null;

        }

        void SetDefaultOptions()
        {
            string exeDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);

            InputFile = Path.Combine(exeDir, "..\\assets\\mov\\big_buck_bunny_trailer_iphone.m4v");
            OutputFile = Path.Combine(exeDir, "reencode_output.mp4");

            Console.WriteLine("Using default options: --input " + InputFile + " --output " + OutputFile);
        }

        public bool Prepare(string[] args)
        {
            ResetOptions();

            if (args.Length == 0)
            {
                SetDefaultOptions();
            }
            else
            {
                if (!CommandLine.Parser.Default.ParseArguments(args, this))
                {
                    Console.WriteLine("Syntax error");
                    PrintUsage();
                    Error = true;
                    return false;
                }
            }

            if (Help)
            {
                PrintUsage();
                return false;
            }

            if (!Validate())
            {
                Console.WriteLine("\nRequired options are not set!");
                PrintUsage();
                Error = true;
                return false;
            }

            return true;
        }

        bool Validate()
        {
            bool res = true;

            Console.Write("Input file: ");
            if (InputFile == null)
            {
                Console.WriteLine("[not set]");
                res = false;
            }
            else
            {
                Console.WriteLine(InputFile);
            }

            Console.Write("Output file: ");
            if (InputFile == null)
            {
                Console.WriteLine("[not set]");
                res = false;
            }
            else
            {
                Console.WriteLine(OutputFile);
            }

            Console.WriteLine("Re-encode audio forced: " + (ReEncodeAudio ? "yes" : "no"));
            Console.WriteLine("Re-encode video forced: " + (ReEncodeVideo ? "yes" : "no"));

            return res;
        }
    }
}
